pkg/controllers/consolidation/controller.go

```diff
@@ -0,0 +1,665 @@
+)
+
+// ProcessCluster is exposed for unit testing purposes
+func (c *Controller) ProcessCluster(ctx context.Context) ProcessResult {
+	candidates := c.candidateNodes(ctx)
+	if len(candidates) == 0 {
+		return ProcessResultNothingToDo
+	}
+
+	emptyNodes := lo.Filter(candidates, func(n candidateNode, _ int) bool { return len(n.pods) == 0 })
+	// first see if there are empty nodes that we can delete immediately, and if so delete them all at once
+	if len(emptyNodes) > 0 {
+		c.performConsolidation(ctx, consolidationAction{
+			oldNodes: lo.Map(emptyNodes, func(n candidateNode, _ int) *v1.Node { return n.Node }),
+			result:   consolidateResultDeleteEmpty,
+		})
+		return ProcessResultConsolidated
+	}
+
+	pdbs, err := NewPDBLimits(ctx, c.kubeClient)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Tracking PodDisruptionBudgets, %s", err)
+		return ProcessResultFailed
+	}
+
+	// the remaining nodes are all non-empty, so we just consolidate the first one that we can
+	sort.Slice(candidates, byNodeDisruptionCost(candidates))
+	for _, node := range candidates {
+		// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+		// scheduling unless its really needed
+		if err = c.canBeTerminated(node, pdbs); err != nil {
+			continue
+		}
+		action := c.nodeConsolidationActions(ctx, node)
+		if action.result == consolidateResultDelete || action.result == consolidateResultShrink {
+			// perform the first consolidation we can since we are looking at nodes in ascending order of disruption cost
+			c.performConsolidation(ctx, action)
+			return ProcessResultConsolidated
+		}
+	}
+	return ProcessResultNothingToDo
+}
+
+// candidateNodes returns nodes that appear to be currently consolidatable based off of their provisioner
+func (c *Controller) candidateNodes(ctx context.Context) []candidateNode {
+	var nodes []candidateNode
+	uninitializedNodeExists := false
+	c.cluster.ForEachNode(func(n *state.Node) bool {
+		// skip any nodes that we can't determine the instance of or for which we don't have consolidation enabled
+		if n.InstanceType == nil || n.Provisioner == nil || n.Provisioner.Spec.Consolidation == nil || !aws.BoolValue(n.Provisioner.Spec.Consolidation.Enabled) {
+			return true
+		}
+		// already found one un-initialized node, so we can skip the rest
+		if uninitializedNodeExists {
+			return true
+		}
+
+		if n.Node.Labels[v1alpha5.LabelNodeInitialized] != "true" {
+			uninitializedNodeExists = true
+			return true
+		}
+		// skip nodes that are annotated as do-not-consolidate
+		if n.Node.Annotations[v1alpha5.DoNotConsolidateNodeAnnotationKey] == "true" {
+			return true
+		}
+
+		// skip nodes that the scheduler thinks will soon have pending pods bound to them
+		if c.cluster.IsNodeNominated(n.Node.Name) {
+			return true
+		}
+
+		pods, err := c.getNodePods(ctx, n.Node.Name)
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Determining node pods, %s", err)
```

Comment:
How spammy is this log?

Reply:
Sure, this is the result of a scale-down that ended up deleting two nodes and replacing one:

```
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:31.214Z	INFO	controller.consolidation	Consolidating via Delete, terminating 1 nodes ip-192-168-166-214.us-west-2.compute.internal/c6g.4xlarge	{"commit": "63daf4b"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:31.248Z	INFO	controller.termination	Cordoned node	{"commit": "63daf4b", "node": "ip-192-168-166-214.us-west-2.compute.internal"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:31.577Z	INFO	controller.termination	Deleted node	{"commit": "63daf4b", "node": "ip-192-168-166-214.us-west-2.compute.internal"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:41.344Z	INFO	controller.consolidation	Consolidating via Delete, terminating 1 nodes ip-192-168-190-133.us-west-2.compute.internal/t4g.xlarge	{"commit": "63daf4b"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:41.385Z	INFO	controller.termination	Cordoned node	{"commit": "63daf4b", "node": "ip-192-168-190-133.us-west-2.compute.internal"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:44:41.678Z	INFO	controller.termination	Deleted node	{"commit": "63daf4b", "node": "ip-192-168-190-133.us-west-2.compute.internal"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:45:12.882Z	INFO	controller.consolidation	Consolidating via Replace, terminating 1 nodes ip-192-168-191-127.us-west-2.compute.internal/c6g.4xlarge and replacing with a node from types a1.2xlarge, t4g.2xlarge, c6g.2xlarge, c7g.2xlarge, t3a.2xlarge and 37 other(s)	{"commit": "63daf4b"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:45:15.281Z	INFO	controller.consolidation.cloudprovider	Launched instance: i-01882e52c50acf61b, hostname: ip-192-168-31-116.us-west-2.compute.internal, type: t4g.2xlarge, zone: us-west-2d, capacityType: on-demand	{"commit": "63daf4b", "provisioner": "default"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:45:15.286Z	INFO	controller.consolidation	Created node with 5 pods requesting {"cpu":"5125m","pods":"10"} from types a1.2xlarge, t4g.2xlarge, c6g.2xlarge, c7g.2xlarge, t3a.2xlarge and 37 other(s)	{"commit": "63daf4b", "provisioner": "default"}
karpenter-68d787db56-6x6q8 controller 2022-07-18T02:46:19.967Z	INFO	controller.termination	Deleted node	{"commit": "63daf4b", "node": "ip-192-168-191-127.us-west-2.compute.internal"}
```

Reply:
No, I can collect another log with debug.

Reply:
```
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:30.491Z	INFO	controller.consolidation	Consolidating via Delete (empty node), terminating 1 nodes ip-192-168-189-114.us-west-2.compute.internal/c6g.8xlarge	{"commit": "ece47b1"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:30.491Z	DEBUG	controller.events	Normal	{"commit": "ece47b1", "object": {"kind":"Node","name":"ip-192-168-189-114.us-west-2.compute.internal","uid":"1e1bc41d-e1c4-402a-a4b0-22ee79d8c882","apiVersion":"v1","resourceVersion":"52764954"}, "reason": "ConsolidateTerminateNode", "message": "Consolidating node via Delete (empty node), terminating 1 nodes ip-192-168-189-114.us-west-2.compute.internal/c6g.8xlarge"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:30.542Z	INFO	controller.termination	Cordoned node	{"commit": "ece47b1", "node": "ip-192-168-189-114.us-west-2.compute.internal"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:30.726Z	INFO	controller.termination	Deleted node	{"commit": "ece47b1", "node": "ip-192-168-189-114.us-west-2.compute.internal"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:40.815Z	INFO	controller.consolidation	Consolidating via Replace, terminating 1 nodes ip-192-168-81-216.us-west-2.compute.internal/c4.8xlarge and replacing with a node from types c6g.8xlarge, c7g.8xlarge, c6a.8xlarge, c6gd.8xlarge, m6g.8xlarge and 10 other(s)	{"commit": "ece47b1"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:41.036Z	DEBUG	controller.consolidation.cloudprovider	Discovered security groups: [sg-05c364fe5ccfc49b4 sg-05f88204faa8cf2d4]	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:41.038Z	DEBUG	controller.consolidation.cloudprovider	Discovered kubernetes version 1.21	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:41.087Z	DEBUG	controller.consolidation.cloudprovider	Discovered ami-064ac13bc8e3c2ee1 for query "/aws/service/eks/optimized-ami/1.21/amazon-linux-2-arm64/recommended/image_id"	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:41.113Z	DEBUG	controller.consolidation.cloudprovider	Discovered ami-00cf63b12c53803a5 for query "/aws/service/eks/optimized-ami/1.21/amazon-linux-2/recommended/image_id"	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:43.191Z	INFO	controller.consolidation.cloudprovider	Launched instance: i-0d136b6c50015d855, hostname: ip-192-168-187-186.us-west-2.compute.internal, type: c6g.8xlarge, zone: us-west-2a, capacityType: on-demand	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:43.250Z	INFO	controller.consolidation	Created node with 22 pods requesting {"cpu":"22125m","pods":"27"} from types c6g.8xlarge, c7g.8xlarge, c6a.8xlarge, c6gd.8xlarge, m6g.8xlarge and 10 other(s)	{"commit": "ece47b1", "provisioner": "default"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:43.250Z	DEBUG	controller.events	Normal	{"commit": "ece47b1", "object": {"kind":"Node","name":"ip-192-168-187-186.us-west-2.compute.internal","uid":"0dbb680e-2c7c-4fe0-aef6-e625e95e1eef","apiVersion":"v1","resourceVersion":"52765122"}, "reason": "ConsolidateLaunchNode", "message": "Launching node for Replace, terminating 1 nodes ip-192-168-81-216.us-west-2.compute.internal/c4.8xlarge and replacing with a node from types c6g.8xlarge, c7g.8xlarge, c6a.8xlarge, c6gd.8xlarge, m6g.8xlarge and 10 other(s)"}
karpenter-68d7cfc48c-rfdg2 controller 2022-07-19T16:17:43.250Z	DEBUG	controller.events	Normal	{"commit": "ece47b1", "object": {"kind":"Node","name":"ip-192-168-187-186.us-west-2.compute.internal","uid":"0dbb680e-2c7c-4fe0-aef6-e625e95e1eef","apiVersion":"v1","resourceVersion":"52765122"}, "reason": "ConsolidateWaiting", "message": "Waiting on readiness to continue consolidation"}
```